core/dmux.go

```diff
@@ -311,7 +378,172 @@ func simpleSetup(size, qsize int, sink Sink) ([]chan interface{}, *sync.WaitGrou
 		go func(index int) {
 			sk := sink.Clone()
 			for msg := range ch[index] {
-				sk.Consume(msg)
+				sk.Consume(msg, math.MaxInt32)
+			}
+			wg.Done()
+		}(i)
+	}
+	return ch, wg
+}
+
+func simpleSetupWithSideline(size, qsize int, sink Sink, sideline Sideline, sidelineImpl interface{}) ([]chan interface{}, *sync.WaitGroup) {
+	wg := new(sync.WaitGroup)
+	wg.Add(size)
+	ch := make([]chan interface{}, size)
+	sidelineChannel := make([]chan ChannelObject, size)
+	sinkChannel := make([]chan ChannelObject, size)
+
+	for i := 0; i < size; i++ {
+		sinkChannel[i] = make(chan ChannelObject, qsize)
+		go func(index int) {
+			sk := sink.Clone()
+			for channelObject := range sinkChannel[index] {
+				log.Printf("Inside Sink channel ")
+				for {
+					consumeError := sk.Consume(channelObject.Msg, sideline.Retries)
+					if consumeError == nil {
+						break
+					}
+					if consumeError != nil && consumeError.Error() == "exceeded retries" {
+						sendToSidelineChannel := ChannelObject{
+							Msg:      channelObject.Msg,
+							Sideline: channelObject.Sideline,
+							Version:  0,
+						}
+						sidelineChannel[index] <- sendToSidelineChannel
+						break
+					}
+					continue
+				}
+			}
+		}(i)
+	}
+
+	for i := 0; i < size; i++ {
+		sidelineChannel[i] = make(chan ChannelObject, qsize)
+		go func(index int) {
+			for channelObject := range sidelineChannel[index] {
+				for {
```

Comment:
Any reason to have this for loop? If its for retry, its not the right approach, since current implemation can lead to an infinite loop with 100% CPU utilization. Better approach would be trough a retry handler which attempts a limited number of times based on the error returned by the underlying function. 

Examples of retry libs

https://github.com/cenkalti/backoff
https://github.com/avast/retry-go